### PR TITLE
[WIP] New ZZ error estimator: SERIAL VERSION

### PR DESCRIPTION
**Pull request by brendankeith**
The purpose of this PR is to provide MFEM with a flux reconstruction error estimator, η, which is both reliable and efficient for arbitrary polynomial orders p. By reliable, we mean that there exists a mesh-independent constant C such that e ≤  C η, where e denotes the true global error. By efficient, we mean that exists another mesh-independent  constant C such that η ≤ C e.

The choice has been made to design this estimator using a local averaging technique first introduced by Zienkiewicz and Zhu in [1,2]. The original work proposes flux reconstruction by solving a least-squares problem which uses global polynomials to fit discrete flux values inside vertex patches. This approach does not generalize well to meshes with an arbitrary number of hanging nodes. Based in part on this drawback, we use a face-patch averaging technique proposed by Bartels and Carstensen [3].

To run: `./ZZ_test -p 1 -rt 0.0 -o 1 -est 1`, `./ZZ_test -p 1 -rt 0.0 -o 2 -est 1`

To create convergence plots: `python error_plot.py`

[1] Zienkiewicz, O.C. & Zhu, J.Z. (1992). The superconvergent patch recovery and a posteriori error estimates. Part 1: The recovery technique. Int. J. Num. Meth. Engng. 33, 1331-1364.

[2] Zienkiewicz, O.C. & Zhu, J.Z. (1992). The superconvergent patch recovery and a posteriori error estimates. Part 2: Error estimates and adaptivity. Int. J. Num. Meth. Engng. 33, 1365-1382.

[3] Bartels, S., & Carstensen, C. (2002). Each averaging technique yields reliable a posteriori error control in FEM on unstructured grids. Part II: Higher order FEM. Math. Comput., 71(239), 971-994.

NOTES:

• This was written foremost for the DRL4AMR LDRD, which currently only requires a serial implementation.  A parallel implementation will come in another PR.

### CODE REVIEW

File: fem/estimators.hpp [hunk 1]
```diff
@@ -202,6 +202,141 @@ class ZienkiewiczZhuEstimator : public AnisotropicErrorEstimator
...
+
+    Zienkiewicz, O.C. and Zhu, J.Z., The superconvergent patch recovery
+    and a posteriori error estimates. Part 1: The recovery technique.
+    Int. J. Num. Meth. Engng. 33, 1331-1364 (1992).
+
+    Zienkiewicz, O.C. and Zhu, J.Z., The superconvergent patch recovery
+    and a posteriori error estimates. Part 2: Error estimates and adaptivity.
+    Int. J. Num. Meth. Engng. 33, 1365-1382 (1992).
+
+    Each averaging technique yields reliable a posteriori error control in
```
Comment (dylan-copeland) [inline fem/estimators.hpp:216]: Authors are missing.

File: fem/gridfunc.cpp [hunk 2]
```diff
@@ -4039,6 +4085,451 @@ double ZZErrorEstimator(BilinearFormIntegrator &blfi,
...
+      //        elements in the face patch
+      int nfdofs = 0;
+      for (int i = 0; i < num_neighbor_elems; i++)
+      {
+         int ielem = neighbor_elems[i];
+         const IntegrationRule *ir = &(IntRules.Get(mesh->GetElementGeometry(ielem),
+                                                    flux_order));
+         int num_integration_pts = ir->GetNPoints();
+
+         udoftrans = ufes->GetElementVDofs(i, udofs);
```
Comment (dylan-copeland) [inline fem/gridfunc.cpp:4343]: Should `i` be `ielem`? Also on the next line.

File: fem/gridfunc.cpp [hunk 3]
```diff
@@ -4039,6 +4085,451 @@ double ZZErrorEstimator(BilinearFormIntegrator &blfi,
...
+         fdoftrans = ffes->GetElementVDofs(i, fdofs);
+         nfdofs += fdofs.Size();
+
+         ufes->GetElementVDofs(ielem, udofs);
```
Comment (dylan-copeland) [inline fem/gridfunc.cpp:4347]: This was called above.

**Comment by tzanio** [comment]
@brendankeith, can we also delete the `zz_test` branch?

**Pull request closed without merging**